src/azul/plugins/repository/tdr_anvil/__init__.py

```diff
@@ -113,7 +115,114 @@ def merge(cls, links: Iterable['Link']) -> 'Link':
+    balance between the size and number of the resulting bundles as well as the
+    degree of overlap between them. The implementation of the graph traversal is
+    tightly coupled to this choice, and switching to a different entity type
+    would require re-implementing much of the Plugin code. Primary bundles
+    consist of at least one biosample (the bundle entity), exactly one dataset,
+    and zero or more other entities of assorted types.
+
+    Some snapshots include file entities that lack any foreign keys that
+    associate the file with any other entity. To ensure that these "orphaned"
+    files are indexed, they are also used as bundle entities. As with primary
+    bundles, the creation of these supplementary bundles depends on a
+    specifically tailored traversal implementation. Supplementary bundles always
+    consist of exactly two entities: one file (the bundle entity) and one
+    dataset.
+    """
+    primary: EntityType = 'biosample'
+    supplementary: EntityType = 'file'
+
+
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class AnvilBundleFQID(TDRBundleFQID):
+    entity_type: BundleEntityType = attr.ib(converter=BundleEntityType)
+
+    def to_json(self) -> SourcedBundleFQIDJSON:
+        return dict(super().to_json(),
+                    entity_type=self.entity_type.value)
+
+
+class TDRAnvilBundle(TDRBundle):
+
+    def add_entity(self,
+                   entity: EntityReference,
+                   version: str,
+                   row: MutableJSON
+                   ) -> None:
+        pk_column = entity.entity_type + '_id'
+        self._add_entity(
+            manifest_entry={
+                'uuid': entity.entity_id,
+                'version': version,
+                'name': f'{entity.entity_type}_{row[pk_column]}',
+                'indexed': True,
+                'crc32': '',
+                'sha256': '',
+                **(
+                    {'drs_path': self._parse_drs_uri(row.get('file_ref'))}
+                    if entity.entity_type == 'file' else {}
+                )
+            },
+            metadata=row
+        )
+
+    def add_links(self,
+                  links: Links,
+                  entities_by_key: Mapping[KeyReference, EntityReference]) -> None:
+        def link_sort_key(link: JSON):
+            return link['activity'] or '', link['inputs'], link['outputs']
+
+        def key_ref_to_entity_ref(key_ref: KeyReference) -> str:
+            return str(entities_by_key[key_ref])
+
+        self._add_entity(
+            manifest_entry={
+                'uuid': self.fqid.uuid,
+                'version': self.fqid.version,
+                'name': 'links',
+                'indexed': True
+            },
+            metadata=sorted((
+                {
+                    'inputs': sorted(map(key_ref_to_entity_ref, link.inputs)),
+                    'activity': None if link.activity is None else key_ref_to_entity_ref(link.activity),
```

Comment:
I'm getting a type warning here: could you please work this patch in? It fixes another type warning, too:

```diff
Index: src/azul/plugins/repository/tdr_anvil/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr_anvil/__init__.py b/src/azul/plugins/repository/tdr_anvil/__init__.py
--- a/src/azul/plugins/repository/tdr_anvil/__init__.py	(revision 01c7ec8c3a4028957f4c746b50ac4dc41c0d185f)
+++ b/src/azul/plugins/repository/tdr_anvil/__init__.py	(date 1682979412879)
@@ -188,6 +188,9 @@
         def key_ref_to_entity_ref(key_ref: KeyReference) -> str:
             return str(entities_by_key[key_ref])
 
+        def optional_key_ref_to_entity_ref(key_ref: Optional[KeyReference]) -> str:
+            return None if key_ref is None else key_ref_to_entity_ref(key_ref)
+
         self._add_entity(
             manifest_entry={
                 'uuid': self.fqid.uuid,
@@ -198,7 +201,7 @@
             metadata=sorted((
                 {
                     'inputs': sorted(map(key_ref_to_entity_ref, link.inputs)),
-                    'activity': None if link.activity is None else key_ref_to_entity_ref(link.activity),
+                    'activity': optional_key_ref_to_entity_ref(link.activity),
                     'outputs': sorted(map(key_ref_to_entity_ref, link.outputs))
                 }
                 for link in links
@@ -417,7 +420,7 @@
         return result
 
     def _simplify_links(self, links: Links) -> None:
-        grouped_links = defaultdict(set)
+        grouped_links: Mapping[KeyReference, Links] = defaultdict(set)
         for link in links:
             grouped_links[link.activity].add(link)
         for activity, convergent_links in grouped_links.items():
```